### PR TITLE
Add MXNet backend - CNN

### PR DESCRIPTION
**Pull request by sandeep-krishnamurthy**
Hello @fchollet  and Keras Team,

In this PR, we are adding the first iteration of MXNet backend implementation for Keras. Request your help in review and merge for this PR to enable Keras community to have MXNet as one of the backends for their Deep Learning tasks.

**Status of this PR**
1. Supports CNN use cases only. (2D only).
2. Supports CPU, 1 GPU, Multi-GPU for large-scale training.
3. Backend Operators - 123 out of 143 implemented.
4. Keras Tests - 375 passed, 197 skipped in 88.88 seconds.

**Not yet supported**
1. Conv1D and 3D use cases.
2. separable_conv, depthwise_conv.
3. RNN use cases.
4. Sparse tensor support.
5. update_* operations.
6. cumsum, cumprod, stack operators

**Known Issues/Things to refactor**
1. MXNet do not support update_* (eg: update_add) operations yet. Hence, MXNet backend inherits Model class and extends it to work with MXNet’s Module APIs. 
2. MXNet backend uses native mxnet batch norm directly due to missing batch_update and add_update functionality.
3. MXNet backend inherits and extends optimizer class due to missing update functions in MXNet.
4. in_topk operator return 0/1 instead of Boolean True/False due to MXNet not supporting Boolean yet.
5. MXNet do not have yet support to optimizers - Adamax, NAdam, Adam_AMSGrad
I have Github issues created to track all above mentioned work and also list of all skipped/failing test cases here - https://github.com/deep-learning-tools/keras/issues. 

Wanted to take incremental approach and get into hands of the keras community to enable more contributions, early testing and suggestions for improvements.

Many thanks to all the contributors for this work: @jiajiechen and contributors of Keras1.2 + MXNet work - @piiswrong, @kevinthesun, @howardsu, @yajiedesign

Thanks,
Sandeep

### CODE REVIEW

**Comment by ahundt** [comment]
This seems great!

I only glanced through, but isn't there a way to bulk skip tests in a whole file? I think there are some unit test files where every function is skipped. A bulk skip at the start would make it much simpler.

File: keras/utils/test_utils.py [hunk 1]
```diff
@@ -92,6 +92,7 @@ def layer_test(layer_cls, kwargs={}, input_shape=None, input_dtype=None,
     # check shape inference
     model = Model(x, y)
     expected_output_shape = layer.compute_output_shape(input_shape)
+    model.compile('rmsprop', 'mse')
```
Comment (fchollet) [inline keras/utils/test_utils.py]: The Keras API does not require `compile` before calling `predict`, because compile merely configures training and is not related to inference. If MXNet requires it, that's a bug and it should be fixed.

File: keras/optimizers.py [hunk 2]
```diff
@@ -15,6 +15,8 @@
 
 if K.backend() == 'tensorflow':
     import tensorflow as tf
+elif K.backend() == 'mxnet':
+    import mxnet as mx
```
Comment (fchollet) [inline keras/optimizers.py]: Doesn't appear to be used

File: keras/optimizers.py [hunk 3]
```diff
@@ -683,6 +685,10 @@ def from_config(self, config):
         raise NotImplementedError
 
 
+if K.backend() == 'mxnet':
+    SGD, Adagrad, Adadelta, Adam, RMSprop = K.get_optimizers()
```
Comment (fchollet) [inline keras/optimizers.py]: The optimizers as implemented in this file should work with MXNet.  Don't replace them with native classes. For instance, current optimizers are shared across TF/Theano/CNTK even though TF has native optimizers.

File: keras/layers/normalization.py [hunk 4]
```diff
@@ -88,6 +87,11 @@ def __init__(self,
         self.beta_constraint = constraints.get(beta_constraint)
         self.gamma_constraint = constraints.get(gamma_constraint)
 
+        if axis == -1 and K.image_data_format() == 'channels_first':
+            self.axis = 1
+        else:
+            self.axis = axis
```
Comment (fchollet) [inline keras/layers/normalization.py]: "-1" means "last axis", not "feature axis". Please revert.

File: keras/layers/normalization.py [hunk 5]
```diff
@@ -128,6 +132,14 @@ def build(self, input_shape):
         self.built = True
 
     def call(self, inputs, training=None):
+
+        # MXNet Backend calculates moving mean and moving variance in batch_norm operation directly.
+        # Hence, calling native mxnet batchnorm.
+        # This is functional for now, however, needs to be revisited to do it in native Keras way.
+        if K.backend() == 'mxnet':
+            return K.mxnet_batchnorm(inputs, self.gamma, self.beta, self.moving_mean, self.moving_variance,
+                                     self.momentum, axis=self.axis, epsilon=self.epsilon)
```
Comment (fchollet) [inline keras/layers/normalization.py]: I don't understand how this can possibly work. BN has a different behavior in inference and training, but it doesn't appear to be reflected here.

What is the problem with using just the backend ops to implement BN?

File: keras/engine/training.py [hunk 6]
```diff
@@ -2503,3 +2503,8 @@ def predict_generator(self, generator, steps=None,
             return [out[0] for out in all_outs]
         else:
             return [np.concatenate(out) for out in all_outs]
+
+# We need to overload the Keras Model class to handle MXNet model building activities.
+# get_model() returns the reference to MXNetModel that inherits and extends keras.engine.Model.
+if K.backend() == 'mxnet':
+    Model = K.get_model()
```
Comment (fchollet) [inline keras/engine/training.py]: This breaks the general philosophy of `keras-team/keras`, which is to have a shared codebase across different backends (which is possible as long as each backend implements the Keras backend API). Why do you think this  is necessary?

We've managed to do it across Theano and TF without much issues, even they work quite differently (e.g. sessions, graphs, etc). Surely you can manage any custom functionality you need at the level of `K.Function()`.

**Comment by fchollet** [comment]
> MXNet do not support update_* (eg: update_add) operations yet

Neither does Theano. They're implemented differently in Theano (not as ops, but as tuples that get assigned during `Function` calls). Similarly you could implement them as deferred-execution functions.

**Comment by fchollet** [comment]
> RNN use cases.

That's a massive limitation. The others are more acceptable (separable convs, 1D convs, 3D convs). Note that you could easily support 1D convs by implementing them as 2D convs.

What's the timeline for you to support RNNs?

File: keras/backend/theano_backend.py [hunk 7]
```diff
@@ -1614,7 +1614,7 @@ def dropout(x, level, noise_shape=None, seed=None):
         seed: random seed to ensure determinism.
     """
     if level < 0. or level >= 1:
-        raise ValueError('Dropout level must be in interval [0, 1[.')
+        raise ValueError('Dropout level must be in interval [0, 1].')
```
[the remark this answers is not in the document]
Comment (sandeep-krishnamurthy) [reply keras/backend/theano_backend.py]: Fixed.

File: keras/layers/convolutional.py [hunk 8]
```diff
@@ -129,7 +129,13 @@ def build(self, input_shape):
             raise ValueError('The channel dimension of the inputs '
                              'should be defined. Found `None`.')
         input_dim = input_shape[channel_axis]
-        kernel_shape = self.kernel_size + (input_dim, self.filters)
+
+        if self.data_format == 'channels_first':
+            kernel_shape = (self.filters, input_dim) + self.kernel_size
+        else:
+            kernel_shape = self.kernel_size + (input_dim, self.filters)
```
[the remark this answers is not in the document]
Comment (sandeep-krishnamurthy) [reply keras/layers/convolutional.py]: Reverted.

But, the major concern here would be a massive performance hit due to extra transpose operation on kernels for backends with "channels_last" format.

Is there any way to avoid it? Say for example, recommending channels_last format for MXNet backend in keras.json config?

File: keras/layers/normalization.py (hunk 4 above)
Comment (sandeep-krishnamurthy) [reply keras/layers/normalization.py]: Reverted.

But this again assumes feature axis to be last by default?
Same request as above, is there any way to allow flexibility by recommending 'channels_last' format for MXNet backend? Thereby standardizing the usage?

File: keras/optimizers.py (hunk 2 above)
Comment (sandeep-krishnamurthy) [reply keras/optimizers.py]: Fixed.

File: keras/layers/convolutional.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (fchollet) [reply keras/layers/convolutional.py]: The same concern exists for Theano with `data_format='channels_last'`.

I do not know the extent of the performance hit. My guess is that it is small. I note that [native TF ops](https://www.tensorflow.org/api_docs/python/tf/nn/conv2d) use the same kernel shape for both data formats, which indicates that performance optimization can probably be handled at the backend level.

File: keras/layers/normalization.py (hunk 5 above)
Comment (sandeep-krishnamurthy) [reply keras/layers/normalization.py]: In the backend, for executing the graph with MXNet module (Executor), a flag is set to indicate whether the stage is training or inference. That was managing different behavior required.

File: keras/optimizers.py (hunk 3 above)
Comment (sandeep-krishnamurthy) [reply keras/optimizers.py]: I agree and ideally, it should be all Keras engine code that manages the backend. I was thinking of getting into hands of users, get early feedback, and create a Github issue to track these and fast follow up after this PR to work on removing MXNet specific things.
Again, this requires update_* mechanisms. I will go through Theano way of handling it and get it implemented for MXNet backend.

**Comment by sandeep-krishnamurthy** [comment]
Thank you @fchollet and @ahundt for initial comments. I have addressed few of the suggested changes.

> Partial comments for now.
> Please note:
> * use ' as string delimiter
> * don't use \ to break lines
> * pytest allows entire files to be skipped in just one line, you don't need to decorate individual test functions

Done. 

> RNN use cases.

> > That's a massive limitation. The others are more acceptable (separable convs, 1D convs, 3D convs). Note that you could easily support 1D convs by implementing them as 2D convs.

> > What's the timeline for you to support RNNs?

I see at least 4-5 weeks from now to get RNN working with MXNet backend.

> MXNet do not support update_* (eg: update_add) operations yet

> > Neither does Theano. They're implemented differently in Theano (not as ops, but as tuples that get assigned during Function calls). Similarly you could implement them as deferred-execution functions.

Thank you for the pointer. I will read through it and try to implement with MXNet backend.

I have reverted changes in layer/convolutional.py related to Kernel shape change based on data_format.

Working on fix for Predict() API. Will be fixed soon in this PR.

For the following issues, I was thinking if I can create Github issues and work on it immediately after this PR:
* Model: Handle MXNet Model logic in K.function() instead of overriding Keras Model class.
* Update_*: Refer to Theano backend and implement update ops for MXNet.
* BatchNorm: MXNet native batch norm -> Keras batch norm
* Optimizers: Once we have update APIs implemented, move all MXNet native optimizer to Keras optimizer.

The reason I am thinking of handling above in upcoming PRs is that all the custom logic is protected under MXNet backend and do not break any other backends. Users can use MXNet backend and provide early feedback. We can possibly get more contributors.

Please do let me know your suggestions on how to proceed. Also, any suggestion on this should be in a branch or contrib till RNN and above listed functionality are implemented?

File: keras/utils/test_utils.py (hunk 1 above)
Comment (sandeep-krishnamurthy) [reply keras/utils/test_utils.py]: This is fixed.

**Comment by datumbox** [comment]
Great work adding MXnet in Keras guys! 

One question, does multi-gpu training/prediction works out of the box?

**Comment by roywei** [comment]
@datumbox Yes, you need to specify in model by passing context:
`gpu_list = ["gpu(0)", "gpu(1)"]`
`model.compile(loss, optimizer, metrics, context=gpu_list)`

**Comment by datumbox** [comment]
@roywei Sounds great! 

I can modify multi_gpu_utils to add support for MXNet backend if you want, but if I open a separate PR it has to be merged after this one. @fchollet Let me know if you want that at this point.

**Comment by sandeep-krishnamurthy** [comment]
Thank you everyone for your code review, feedback and interest in MXNet backend for Keras. Thank you @fchollet for all your support and suggestion in this work. I am closing this PR for now, as we have one major incompatibility between Keras core and MXNet:
1. MXNet does not support shared tensor. Requires significant changes in core engine and parameter server. Symbolic optimizers are also not supported currently in MXNet.

However, since for the Keras user, Keras interface is same and around 128 out of 143 operators are supported with MXNet backend, for the benefit of Keras community, Keras-MXNet will be released for now in the fork repo.
1. Repo - https://github.com/awslabs/keras-apache-mxnet
2. Installation [https://github.com/awslabs/keras-apache-mxnet/wiki/Installation]f

Please try Keras with MXNet backend.
```
pip install keras-mxnet --pre
```
